### ISSUE TITLE
"Not a valid index for sequence alignment"

### ISSUE THREAD

**Issue by sacdallago** [debbiemarkslab/EVcouplings#175]
Seems related to #151 ( 
@aggreen ?) somehow, getting it from a run via the web interface.

```
Traceback (most recent call last):
  File "/n/groups/marks/software/anaconda_o2/envs/evcouplings_backend_production/lib/python3.5/site-packages/evcouplings/utils/pipeline.py", line 387, in execute_wrapped
    outcfg = execute(**config)
  File "/n/groups/marks/software/anaconda_o2/envs/evcouplings_backend_production/lib/python3.5/site-packages/evcouplings/utils/pipeline.py", line 183, in execute
    outcfg = runner(**incfg)
  File "/n/groups/marks/software/anaconda_o2/envs/evcouplings_backend_production/lib/python3.5/site-packages/evcouplings/compare/protocol.py", line 1034, in run
    return PROTOCOLS[kwargs["protocol"]](**kwargs)
  File "/n/groups/marks/software/anaconda_o2/envs/evcouplings_backend_production/lib/python3.5/site-packages/evcouplings/compare/protocol.py", line 499, in standard
    "prefix": aux_prefix,
  File "/n/groups/marks/software/anaconda_o2/envs/evcouplings_backend_production/lib/python3.5/site-packages/evcouplings/compare/protocol.py", line 101, in _identify_structures
    **kwargs
  File "/n/groups/marks/software/anaconda_o2/envs/evcouplings_backend_production/lib/python3.5/site-packages/evcouplings/compare/sifts.py", line 715, in by_alignment
    map_j, map_j_df = _create_mapping(r)
  File "/n/groups/marks/software/anaconda_o2/envs/evcouplings_backend_production/lib/python3.5/site-packages/evcouplings/compare/sifts.py", line 644, in _create_mapping
    ali[r["alignment_id"]], r["alignment_start"], r["alignment_end"]
  File "/n/groups/marks/software/anaconda_o2/envs/evcouplings_backend_production/lib/python3.5/site-packages/evcouplings/align/alignment.py", line 647, in __getitem__
    "Not a valid index for sequence alignment: {}".format(index)
KeyError: 'Not a valid index for sequence alignment: sp|O00212|RHOD_HUMAN/19-58'
```

check `/n/scratch2/105280abb7b4c88bbc1e7db8761d45e8`. I am 80% sure that the version running on o2 for the web server is the latest code on develop merged with my web branch. I'm keeping this here as question to see if this was a known, resolved issue. If not, might be worth checking out.

**Comment by thomashopf** [comment]
No, this is a different issue.

The problem is that the query sequence identifier used for the run is a "full" Uniprot identifier: `sp|P01112|RASH_HUMAN`. Once jackhmmer finds a sequence with the same name in the database (sp|P01112|RASH_HUMAN) it starts making unique sequence identifiers in the .sto alignment by prefixing with numbers, but annoyingly enough it doesn't do that in the domtblout. Once the identifiers from the domtblout are used to access the alignment, this creates the KeyError.

Short-term solution: Don't use a full Uniprot header for the search but just P01112 or RASH_HUMAN as name.

Long-term solution: Actively avoid the problem, e.g. by adding some prefix / suffix to the target sequence identifier before searching.

**Comment by sacdallago** [comment]
I see. So the short-term solution is a hacky change in the pipeline (rather than the submission process) and the long-term solution is a less hacky change in the pipeline or external software?

I'm just trying to understand if this is something that generates at all from _how_ I submit the job (which as I understand it, it isn't). If it is so, this def. has much less priority for me than a couple other things.

Thanks!

**Comment by thomashopf** [comment]
No the short-term solution is in fact w.r.t. *how* one submits, to avoid jackhmmer running into the problematic behaviour in the first place: If one doesn't use the full header from the Uniprot sequence file (e.g. `sp|P01112|RASH_HUMAN` - this breaks) , the identifier clash problem doesn't occur. Because the pipeline itself replaces the header line with the user-defined sequence identifier (e.g. just `RASH_HUMAN`), this problem typically never occurs unless one explicitly inputs such a problematic identifier as `sequence_id`..

The quick solution for the webserver would be to ensure that sth. like `sp|P01112|RASH_HUMAN` gets defused. E.g. if the user inputs `P01112` into the input form, just use that - and if they input a full FASTA entry that has `sp|P01112|RASH_HUMAN`, replace the pipes with "_", ...

The only really clean long-term solution would be if jackhmmer behaved consistently. Everything else is a hack.

**Comment by sacdallago** [comment]
Hmm. Okay. There are a few things which are confusing, still. This is a run that you submitted? If so, what was your input in the web interface like? This is important, here's why:

1. If you input `RASH_HUMAN` or `P01112`, what will be submitted to the backend is always and only `P01112`
2. If you input a FASTA-like sequence BUT you specify a correct uniprot header with a correct uniprot reference, the web server will fetch that sequence from uniprot and **IGNORE** whatever the user inputs. This is a design choice (we can argue about weather this is good or bad, but for now this is how it is and this is why I still can't wrap my head around this error, if, as you say, it's a submission error). So, again, if you were to input:

```
>sp|P01112|RASH_HUMAN
AAAAAAAA
```

in the web interface, this is the exact same as inputting `P01112` or `RASH_HUMAN`, and the uniprot sequence will be used, and `P01112` (plus an eventual selected region) will be sent to the backend for the job, **not** `AAAAAAAA`. I tried to make this clear by adding the following on the submission page:
![screen shot 2018-06-18 at 1 11 00 pm](https://user-images.githubusercontent.com/6240943/41551179-2089e6b4-72f9-11e8-9222-cbb5e8ff07a1.png)

w.r.t. to a "real" fasta-like input:
![screen shot 2018-06-18 at 1 11 13 pm](https://user-images.githubusercontent.com/6240943/41551200-2d56ddf2-72f9-11e8-9f5f-1b4d3af9ab5c.png)


Again: let's not deviate from the discussion here, the design choice weather good or bad is a separate issue: I'm trying to understand: What was the input on the web interface?

**Comment by thomashopf** [comment]
Don't know if this was one of my runs or not... could be. When I played around with the server, I think I only input "P01112", not a sequence + header.

Anyways, what fundamentally causes the problem is `sequence_id: sp|P01112|RASH_HUMAN` in the global section of the config file, which must have been put in there by the backend in some way or the other. If this was `sequence_id: RASH_HUMAN`or `sequence_id: P01112` the crash wouldn't occur.

(Note that the content of the FASTA file if downloaded from Uniprot by the pipeline, or supplied within `sequence_file` doesn't matter, since the header line gets replaced by `sequence_id` (+region))

So bottom line: Avoid putting something like `sp|P01112|RASH_HUMAN` as value of `sequence_id` at all cost for now... not a terrible restriction to impose.

(And I just had another idea for getting around this problem this within the pipeline relatively elegantly eventually, without having to rely on fixing jackhmmer or patching user inputs).

**Comment by sacdallago** [comment]
> So bottom line: Avoid putting something like `sp|P01112|RASH_HUMAN` as value of sequence_id at all cost for now... not a terrible restriction to impose.

absolutely not terrible. 😄  I wonder at all **how** this may have happened. As I explained earlier, the web passes either `P01112` or just the sequence, like 'ABC...' and no header (it should apply a default header, something like `>sequence`.

I'll double check on my side then.

For the fix: would you mind opening an issue here about it and describing exactly what you would want to see? Or even in this issue. Thanks!

**Comment by thomashopf** [comment]
I've decided to make the pipeline robust w.r.t to that case when I get round to it.. the changes are minimal. At that point it won't matter what the webserver puts in.

**Comment by sacdallago** [comment]
Thanks @thomashopf :) makes my life a lot easier ❤️ I'll close my issue then

**Comment by thomashopf** [comment]
This is fixed now for the compare stage (unless someone willfully chooses an identifier to break it) as soon as you regenerate the SIFTS sequence database file with the updated code.

There is a possibility of creating identifier clashes in the align stage too if you put in the wrong identifiers (much less likely though, since the input region will most often differ from the aligned region in the self-hit), but fixing this in full generality is more involved and not worth it.
For these cases , the code now raises an explicit exception if the alignment parser detects HMMER renaming the sequences so it doesn't break in a weird way further downstream.

What this means for server: replace any pipe symbol with an underscore (or some other character) before you put the value for sequence_id in the config file. That should prevent any further problems against the Uniprot/Uniref databases we typically search.

**Comment by sacdallago** [comment]
Whops sorry :D
